File: apps/web/lib/api/competitors.ts
/**
 * Competitor intelligence, over HTTP.
 *
 * 🔴 THE MATRIX IS NOT A RECIPE, AND THIS MODULE MUST NOT LET IT LOOK LIKE ONE.
 *
 * The specification forbids presenting an inferred competitor composition as a
 * known or verified formula. So `compositionMatrixSchema` carries a
 * `disclaimer` the SERVER supplies and the screen renders — not a sentence the
 * screen remembers to add, because a screen that forgets it would be doing the
 * one thing the specification rules out.
 *
 * ⚠️ CONCENTRATIONS ARE STRINGS. `NUMERIC(7,4)` in PostgreSQL, and the server
 * stringifies them at the boundary. Parsing to `number` here would reintroduce
 * the float CLAUDE.md §5 forbids on a controlled record, and would render
 * "10.0000" — a range disclosed to four decimal places — as "10".
 */

import { z } from "zod";

import { apiRequest, type ApiCredentials } from "./client";
import { API_BASE_URL } from "./config";

export const competitorProductSchema = z.object({
  id: z.string(),
  manufacturer: z.string(),
  product_name: z.string(),
  product_code: z.string().nullable(),
  market_segment: z.string().nullable(),
  project_id: z.string().nullable(),
  created_at: z.string(),
  document_count: z.number(),
  evidence_count: z.number(),
});

/**
 * How a claim is known. SEPARATE from the document's type: a person reading a
 * tin is making an observation, not an inference, and the two must not collapse
 * into one field.
 */
export const EVIDENCE_SOURCES = [
  { id: "document", label: "A document on file", needsDocument: true },
  { id: "manual_observation", label: "Read from the product myself", needsDocument: false },
  { id: "laboratory", label: "Our own laboratory result", needsDocument: false },
  { id: "literature", label: "Published literature", needsDocument: false },
  { id: "patent", label: "A patent", needsDocument: false },
  { id: "inference", label: "Inferred from the above", needsDocument: false },
  { id: "model", label: "Model hypothesis", needsDocument: false },
] as const;

/** The A–X ranking from the research source document. */
export const EVIDENCE_GRADES = [
  { id: "A", label: "A — validated internal evidence, a standard, or manufacturer documentation" },
  { id: "B", label: "B — peer-reviewed literature, a patent, or a recognised institution" },
  { id: "C", label: "C — supplier literature or a conference paper" },
  { id: "D", label: "D — a general web source" },
  { id: "X", label: "X — unverified or unreliable" },
] as const;

export const evidenceRowSchema = z.object({
  id: z.string(),
  component_name: z.string(),
  cas_number: z.string().nullable(),
  component_function: z.string().nullable(),
  // Strings. See the header.
  concentration_low: z.string().nullable(),
  concentration_high: z.string().nullable(),
  is_balance: z.boolean(),
  evidence_source: z.string(),
  evidence_grade: z.string(),
  confidence: z.enum(["verified", "supported", "probable", "possible", "unknown"]),
  source_locator: z.string().nullable(),
  rationale: z.string().nullable(),
  verified_at: z.string().nullable(),
  source_document_id: z.string().nullable(),
  sample_id: z.string().nullable(),
  test_id: z.string().nullable(),
  source_document_title: z.string().nullable(),
  source_document_type: z.string().nullable(),
});

export const compositionMatrixSchema = z.object({
  rows: z.array(evidenceRowSchema),
  summary: z.record(z.string(), z.number()),
  // 🔴 SUPPLIED BY THE SERVER AND RENDERED VERBATIM.
  disclaimer: z.string(),
});

export const competitorDocumentSchema = z.object({
  id: z.string(),
  document_type: z.string(),
  title: z.string(),
  content_type: z.string().nullable(),
  byte_size: z.number().nullable(),
  issued_on: z.string().nullable(),
  expires_on: z.string().nullable(),
  created_at: z.string(),
});

export const competitorSampleSchema = z.object({
  id: z.string(),
  sample_reference: z.string(),
  acquired_on: z.string().nullable(),
  batch_marking: z.string().nullable(),
  observations: z.string().nullable(),
  registered_by: z.string(),
  created_at: z.string(),
  // A count, and it is named as one. `has_root_cause` was a column whose name
  // asked a yes/no question and whose value was a number (2026-08-27); a field
  // called `evidence_count` can only be read as the number it is.
  evidence_count: z.number(),
});

export const competitorBenchmarkSchema = z.object({
  id: z.string(),
  attribute: z.string(),
  competitor_value: z.string().nullable(),
  our_value: z.string().nullable(),
  gap_summary: z.string(),
  project_id: z.string(),
  project_name: z.string().nullable(),
  project_code: z.string().nullable(),
  formula_version_id: z.string().nullable(),
  test_id: z.string().nullable(),
  recorded_by: z.string(),
  created_at: z.string(),
  // 🔴 NO DISPOSITION FIELD, DELIBERATELY. Testing owns GREEN/YELLOW/RED and
  // the server does not send one; a colour invented here would be a second
  // answer to a question Testing already answers.
});

export type CompetitorProduct = z.infer<typeof competitorProductSchema>;
export type EvidenceRow = z.infer<typeof evidenceRowSchema>;
export type CompositionMatrix = z.infer<typeof compositionMatrixSchema>;
export type CompetitorDocument = z.infer<typeof competitorDocumentSchema>;
export type CompetitorSample = z.infer<typeof competitorSampleSchema>;
export type CompetitorBenchmark = z.infer<typeof competitorBenchmarkSchema>;

export interface ProductRequest {
  readonly manufacturer: string;
  readonly product_name: string;
  readonly product_code?: string;
  readonly market_segment?: string;
  readonly notes?: string;
}

export interface EvidenceRequest {
  readonly component_name: string;
  readonly evidence_source: string;
  readonly evidence_grade: string;
  readonly cas_number?: string;
  readonly component_function?: string;
  readonly concentration_low?: string;
  readonly concentration_high?: string;
  readonly is_balance?: boolean;
  readonly source_document_id?: string;
  // 🔴 THE SERVER HAS ALWAYS ACCEPTED THIS AND NOTHING EVER SENT IT.
  // `manual_observation` means somebody read a physical tin; without naming
  // WHICH tin, the claim cannot be re-checked and the grade is unearned.
  readonly sample_id?: string;
  readonly source_locator?: string;
  readonly rationale?: string;
}

export interface SampleRequest {
  readonly sample_reference: string;
  readonly acquired_on?: string;
  readonly batch_marking?: string;
  readonly observations?: string;
}

export interface BenchmarkRequest {
  readonly project_id: string;
  readonly attribute: string;
  readonly gap_summary: string;
  readonly competitor_value?: string;
  readonly our_value?: string;
  readonly formula_version_id?: string;
  readonly test_id?: string;
}

export function fetchCompetitorProducts(
  credentials: ApiCredentials,
  signal?: AbortSignal,
): Promise<CompetitorProduct[]> {
  return apiRequest({ path: "/api/competitors", credentials, signal }, (payload) =>
    z.array(competitorProductSchema).parse(payload),
  );
}

export function fetchCompositionMatrix(
  credentials: ApiCredentials,
  productId: string,
  signal?: AbortSignal,
): Promise<CompositionMatrix> {
  return apiRequest(
    { path: `/api/competitors/${productId}/composition`, credentials, signal },
    (payload) => compositionMatrixSchema.parse(payload),
  );
}

export function fetchCompetitorDocuments(
  credentials: ApiCredentials,
  productId: string,
  signal?: AbortSignal,
): Promise<CompetitorDocument[]> {
  return apiRequest(
    { path: `/api/competitors/${productId}/documents`, credentials, signal },
    (payload) => z.array(competitorDocumentSchema).parse(payload),
  );
}

export function registerCompetitorProduct(
  credentials: ApiCredentials,
  request: ProductRequest,
): Promise<{ id: string }> {
  return apiRequest(
    { path: "/api/competitors", method: "POST", credentials, body: request },
    (payload) => z.object({ id: z.string() }).parse(payload),
  );
}

export function recordCompetitorEvidence(
  credentials: ApiCredentials,
  productId: string,
  request: EvidenceRequest,
): Promise<{ id: string; confidence: string }> {
  return apiRequest(
    {
      path: `/api/competitors/${productId}/evidence`,
      method: "POST",
      credentials,
      body: request,
    },
    (payload) => z.object({ id: z.string(), confidence: z.string() }).parse(payload),
  );
}

export function gradeCompetitorEvidence(
  credentials: ApiCredentials,
  evidenceId: string,
  confidence: string,
): Promise<{ id: string; confidence: string }> {
  return apiRequest(
    {
      path: `/api/competitors/evidence/${evidenceId}/grade`,
      method: "POST",
      credentials,
      body: { confidence },
    },
    (payload) => z.object({ id: z.string(), confidence: z.string() }).parse(payload),
  );
}

/**
 * Upload a label or a product photograph.
 *
 * 🔴 MULTIPART, AND IT DOES NOT GO THROUGH `apiRequest`.
 *
 * `apiRequest` JSON-encodes its body and sets `Content-Type: application/json`.
 * A file needs `FormData` and the browser's own boundary — setting the header
 * by hand omits the boundary and the server cannot parse the parts. So this is
 * a deliberate exception, and it still carries the same credentials and
 * organization header every other call does.
 */
export async function uploadCompetitorDocument(
  credentials: ApiCredentials,
  productId: string,
  file: File,
  documentType: string,
  title: string,
): Promise<{ id: string }> {
  const body = new FormData();
  body.append("file", file);
  body.append("document_type", documentType);
  body.append("title", title);

  if (API_BASE_URL === null) {
    // The same absence `apiRequest` distinguishes: this build has nothing to
    // call, which is not a failure and must not read as one.
    throw new Error("this build is not pointed at an API, so nothing can be uploaded");
  }
  const response = await fetch(`${API_BASE_URL}/api/competitors/${productId}/documents`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${credentials.token}`,
      "X-Organization-Id": credentials.organizationId,
      // NO Content-Type: the browser sets it, WITH the multipart boundary.
    },
    body,
  });

  if (!response.ok) {
    let detail = `the upload failed (${response.status})`;
    try {
      const parsed = (await response.json()) as { detail?: unknown };
      if (typeof parsed.detail === "string") detail = parsed.detail;
    } catch {
      // A non-JSON error body. The status alone is what we have.
    }
    throw new Error(detail);
  }
  return z.object({ id: z.string() }).parse(await response.json());
}


export function fetchCompetitorSamples(
  credentials: ApiCredentials,
  productId: string,
  signal?: AbortSignal,
): Promise<CompetitorSample[]> {
  return apiRequest(
    { path: `/api/competitors/${productId}/samples`, credentials, signal },
    (payload) => z.array(competitorSampleSchema).parse(payload),
  );
}

export function fetchCompetitorBenchmarks(
  credentials: ApiCredentials,
  productId: string,
  signal?: AbortSignal,
): Promise<CompetitorBenchmark[]> {
  return apiRequest(
    { path: `/api/competitors/${productId}/benchmarks`, credentials, signal },
    (payload) => z.array(competitorBenchmarkSchema).parse(payload),
  );
}

export function registerCompetitorSample(
  credentials: ApiCredentials,
  productId: string,
  request: SampleRequest,
): Promise<{ id: string }> {
  return apiRequest(
    {
      path: `/api/competitors/${productId}/samples`,
      method: "POST",
      credentials,
      body: request,
    },
    (payload) => z.object({ id: z.string() }).parse(payload),
  );
}

export function recordCompetitorBenchmark(
  credentials: ApiCredentials,
  productId: string,
  request: BenchmarkRequest,
): Promise<{ id: string }> {
  return apiRequest(
    {
      path: `/api/competitors/${productId}/benchmarks`,
      method: "POST",
      credentials,
      body: request,
    },
    (payload) => z.object({ id: z.string() }).parse(payload),
  );
}

File: apps/api/app/api/competitors.py
"""Competitor intelligence, over HTTP.

🔴 THE LABEL UPLOAD GOES THROUGH THE MATERIALS DOCUMENT WRITER.

`POST /api/competitors/{id}/documents` does not store a file. It calls
`materials.store_document` — the same function an SDS goes through — with a
competitor product as the owner instead of a material. So a competitor's label
gets the identical treatment: type validation against the real bytes, a malware
scan whose verdict is recorded, a checksum, an object-storage key, and the
supersession rules. §14: *"Do not build a second document repository."*

That is also why the status codes below are the same five: they are the
document writer's, not this module's.

🔴 EVIDENCE IS RECORDED AT `possible`. NOTHING HERE CAN CREATE A VERIFIED CLAIM.

`POST .../evidence` takes no confidence. Moving a claim to `verified` is a
separate route requiring `compliance.review_sds`, and the database additionally
refuses unless the named verifier actually holds it. The specification forbids
presenting an inferred competitor recipe as a known formula, and this is the
mechanism rather than the intention.
"""

from __future__ import annotations

import datetime as dt
import uuid
from typing import Any

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile, status
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.core.documents import get_object_store, get_scanner
from app.core.file_types import FileTypeRejectedError
from app.core.malware import (
    MalwareFoundError,
    MalwareScannerPort,
    MalwareScanUnavailableError,
)
from app.core.object_storage import ObjectStorageError, ObjectStoragePort
from app.core.security import Principal, get_db, require_permission
from app.core.tenancy import CrossTenantReferenceError
from app.domains.competitor_intelligence.service import (
    CompetitorError,
    CompetitorNotFoundError,
    CompetitorStateError,
    EvidenceInput,
    adopt_public_product,
    composition_matrix,
    list_benchmarks,
    list_products,
    list_samples,
    record_benchmark,
    record_evidence,
    register_product,
    register_sample,
    verify_evidence,
)
from app.domains.materials.service import (
    DocumentInput,
    MaterialInvalidError,
    MaterialNotFoundError,
    list_competitor_documents,
    store_document,
)

router = APIRouter()

__all__ = ["router"]


class ProductCreate(BaseModel):
    manufacturer: str = Field(min_length=1, max_length=300)
    product_name: str = Field(min_length=1, max_length=300)
    product_code: str | None = Field(default=None, max_length=100)
    market_segment: str | None = Field(default=None, max_length=200)
    # Optional by design: most competitor products are public and belong to the
    # organization rather than to one project.
    project_id: uuid.UUID | None = None
    notes: str | None = Field(default=None, max_length=4000)


class SampleCreate(BaseModel):
    sample_reference: str = Field(min_length=1, max_length=100)
    acquired_on: dt.date | None = None
    batch_marking: str | None = Field(default=None, max_length=200)
    observations: str | None = Field(default=None, max_length=4000)


class EvidenceCreate(BaseModel):
    component_name: str = Field(min_length=1, max_length=300)
    # 🔴 NO `confidence` FIELD. A claim is recorded as `possible` and promoted
    # only by a reviewer. Accepting it here would let the caller decide whether
    # their own guess counts as verified.
    evidence_source: str = Field(
        pattern="^(document|manual_observation|laboratory|literature|patent|inference|model)$"
    )
    evidence_grade: str = Field(pattern="^[ABCDX]$")
    cas_number: str | None = Field(default=None, pattern=r"^[0-9]{2,7}-[0-9]{2}-[0-9]$")
    component_function: str | None = Field(default=None, max_length=200)
    # Strings all the way to PostgreSQL's NUMERIC. A float here would round the
    # disclosed range before the database ever saw it (CLAUDE.md §5).
    concentration_low: str | None = Field(default=None, pattern=r"^[0-9]{1,3}(\.[0-9]{1,4})?$")
    concentration_high: str | None = Field(default=None, pattern=r"^[0-9]{1,3}(\.[0-9]{1,4})?$")
    is_balance: bool = False
    source_document_id: uuid.UUID | None = None
    sample_id: uuid.UUID | None = None
    test_id: uuid.UUID | None = None
    source_locator: str | None = Field(default=None, max_length=500)
    rationale: str | None = Field(default=None, max_length=4000)


class EvidenceGrade(BaseModel):
    confidence: str = Field(pattern="^(verified|supported|probable|possible|unknown)$")


class BenchmarkCreate(BaseModel):
    project_id: uuid.UUID
    attribute: str = Field(min_length=1, max_length=200)
    gap_summary: str = Field(min_length=1, max_length=2000)
    competitor_value: str | None = Field(default=None, max_length=200)
    our_value: str | None = Field(default=None, max_length=200)
    formula_version_id: uuid.UUID | None = None
    test_id: uuid.UUID | None = None


def _refuse(exc: CompetitorError) -> HTTPException:
    if isinstance(exc, CompetitorNotFoundError):
        return HTTPException(status.HTTP_404_NOT_FOUND, str(exc))
    if isinstance(exc, CompetitorStateError):
        return HTTPException(status.HTTP_409_CONFLICT, str(exc))
    return HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, str(exc))


@router.get("", summary="Competitor products this caller can reach")
def get_products(
    principal: Principal = Depends(require_permission("material.view")),
    session: Session = Depends(get_db),
    limit: int = Query(default=200, ge=1, le=500),
) -> list[dict[str, Any]]:
    """Gated on `material.view`, not a new `competitor.view`.

    A competitor product is technical reference material of the same kind as a
    raw material, and the roles that may look at one are the roles that need
    the other. Minting a permission nobody holds is the defect this project has
    caught five times, and the catalogue already carries 29 permissions with no
    enforcement point.
    """
    return list_products(session, organization_id=principal.organization_id, limit=limit)


@router.post("", status_code=status.HTTP_201_CREATED, summary="Register a competitor product")
def post_product(
    payload: ProductCreate,
    principal: Principal = Depends(require_permission("material.edit")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    try:
        result = register_product(
            session,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            manufacturer=payload.manufacturer,
            product_name=payload.product_name,
            product_code=payload.product_code,
            market_segment=payload.market_segment,
            project_id=payload.project_id,
            notes=payload.notes,
        )
    except CompetitorError as exc:
        raise _refuse(exc) from exc
    except CrossTenantReferenceError as exc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, str(exc)) from exc
    session.commit()
    return result


@router.post(
    "/{competitor_product_id}/documents",
    status_code=status.HTTP_201_CREATED,
    summary="Upload a label, a product photograph, an SDS or literature",
)
async def post_competitor_document(
    competitor_product_id: uuid.UUID,
    file: UploadFile = File(..., description="The label, photograph or document itself."),
    document_type: str = Form(
        ..., description="label | product_image | SDS | TDS | literature | patent | other"
    ),
    title: str = Form(...),
    issued_on: dt.date | None = Form(None),
    expires_on: dt.date | None = Form(None),
    principal: Principal = Depends(require_permission("material.edit")),
    session: Session = Depends(get_db),
    store: ObjectStoragePort = Depends(get_object_store),
    scanner: MalwareScannerPort = Depends(get_scanner),
) -> dict[str, str]:
    """The operator's first entry mode: a photograph of the label, or of the tin.

    🔴 IT DELEGATES TO `materials.store_document`. Not a copy of it, and not a
    second document table — the same writer, with a competitor product as the
    owner. So this upload is type-validated against its real bytes, scanned,
    checksummed and stored exactly as a Safety Data Sheet is, and it appears in
    `materials.usable_documents` under the same five conditions.

    ⚠️ UPLOADING DOES NOT POPULATE THE EVIDENCE MATRIX. There is no automatic
    extraction (chosen 2026-08-28: no OCR dependency, and neither installed
    Ollama model can read an image). The file is stored as the evidence a claim
    can CITE; a person then records what it says, and each claim points back at
    this document and the place in it. Pretending otherwise would put invented
    components on a competitor's product.

    STATUS CODES, each a different statement:
      201 stored, scanned clean, citable as evidence
      400 not an accepted type, or the bytes contradict the name
      404 no such competitor product in this organization
      422 the malware scanner found something
      503 no verdict could be obtained -- NOT "clean"
    """
    data = await file.read()
    try:
        document_id = store_document(
            session,
            competitor_product_id=competitor_product_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            spec=DocumentInput(
                document_type=document_type,
                title=title,
                content_type=file.content_type,
                issued_on=issued_on,
                expires_on=expires_on,
                supersedes_id=None,
            ),
            data=data,
            filename=file.filename or "label",
            store=store,
            scanner=scanner,
        )
    except MaterialNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    except MaterialInvalidError as exc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, str(exc)) from exc
    except FileTypeRejectedError as exc:
        # 400, not 422: 422 is reserved for the scanner's verdict, so a client
        # can tell "send a different file" from "that file was hostile".
        raise HTTPException(status.HTTP_400_BAD_REQUEST, str(exc)) from exc
    except MalwareFoundError as exc:
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_CONTENT,
            f"the uploaded file was identified as malware: {exc.signature}",
        ) from exc
    except MalwareScanUnavailableError as exc:
        # 🔴 503, NEVER 201. "No verdict" is not "clean". Storing an unscanned
        # label as citable evidence is how a hostile upload becomes a record
        # somebody later relies on -- and it would be invisible: every file
        # admitted, nothing in the logs, the control simply absent.
        raise HTTPException(
            status.HTTP_503_SERVICE_UNAVAILABLE,
            f"uploads are unavailable because no malware scan could be performed: {exc}",
        ) from exc
    except ObjectStorageError as exc:
        raise HTTPException(
            status.HTTP_503_SERVICE_UNAVAILABLE, f"the document store is unavailable: {exc}"
        ) from exc
    session.commit()
    return {"id": str(document_id)}


@router.get("/{competitor_product_id}/documents", summary="Labels and documents on file")
def get_competitor_documents(
    competitor_product_id: uuid.UUID,
    principal: Principal = Depends(require_permission("material.view")),
    session: Session = Depends(get_db),
) -> list[dict[str, Any]]:
    """What has been uploaded, so a claim can cite one of them."""
    return list_competitor_documents(
        session,
        organization_id=principal.organization_id,
        competitor_product_id=competitor_product_id,
    )


@router.post(
    "/{competitor_product_id}/samples",
    status_code=status.HTTP_201_CREATED,
    summary="Register a physical sample",
)
def post_sample(
    competitor_product_id: uuid.UUID,
    payload: SampleCreate,
    principal: Principal = Depends(require_permission("material.edit")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    try:
        result = register_sample(
            session,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            competitor_product_id=competitor_product_id,
            sample_reference=payload.sample_reference,
            acquired_on=payload.acquired_on.isoformat() if payload.acquired_on else None,
            batch_marking=payload.batch_marking,
            observations=payload.observations,
        )
    except CompetitorError as exc:
        raise _refuse(exc) from exc
    session.commit()
    return result


@router.get("/{competitor_product_id}/samples", summary="Physical samples on file")
def get_samples(
    competitor_product_id: uuid.UUID,
    principal: Principal = Depends(require_permission("material.view")),
    session: Session = Depends(get_db),
) -> list[dict[str, Any]]:
    """The tins we actually hold, so an observation can say which one it read.

    `material.view` and not `test.view`: a sample is part of the competitor
    record, and the person transcribing a label is doing material work.
    """
    return list_samples(
        session,
        organization_id=principal.organization_id,
        competitor_product_id=competitor_product_id,
    )


@router.get("/{competitor_product_id}/composition", summary="The Composition Evidence Matrix")
def get_composition(
    competitor_product_id: uuid.UUID,
    principal: Principal = Depends(require_permission("material.view")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """A candidate composition where every line says what it rests on.

    🔴 THE RESPONSE CARRIES ITS OWN DISCLAIMER, in the payload rather than left
    to the screen to remember. A client that forgot it would be presenting an
    inferred recipe as a known formula, which the specification forbids
    outright — so the words travel with the data.
    """
    return composition_matrix(
        session,
        organization_id=principal.organization_id,
        competitor_product_id=competitor_product_id,
    )


@router.post(
    "/{competitor_product_id}/evidence",
    status_code=status.HTTP_201_CREATED,
    summary="Record one claim about what the product contains",
)
def post_evidence(
    competitor_product_id: uuid.UUID,
    payload: EvidenceCreate,
    principal: Principal = Depends(require_permission("material.edit")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """Recorded at `possible`. Promotion is a separate, permissioned act."""
    try:
        result = record_evidence(
            session,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            competitor_product_id=competitor_product_id,
            spec=EvidenceInput(
                component_name=payload.component_name,
                evidence_source=payload.evidence_source,
                evidence_grade=payload.evidence_grade,
                cas_number=payload.cas_number,
                component_function=payload.component_function,
                concentration_low=payload.concentration_low,
                concentration_high=payload.concentration_high,
                is_balance=payload.is_balance,
                source_document_id=payload.source_document_id,
                sample_id=payload.sample_id,
                test_id=payload.test_id,
                source_locator=payload.source_locator,
                rationale=payload.rationale,
            ),
        )
    except CompetitorError as exc:
        raise _refuse(exc) from exc
    session.commit()
    return result


@router.post("/evidence/{evidence_id}/grade", summary="Change a claim's confidence")
def post_grade(
    evidence_id: uuid.UUID,
    payload: EvidenceGrade,
    principal: Principal = Depends(require_permission("compliance.review_sds")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """Moving a claim to `verified` is a controlled act.

    Gated on `compliance.review_sds` — the same permission that confirms a
    Safety Data Sheet reading, because both are judgements about whether a
    documented claim may be relied upon. The database independently refuses
    unless the named verifier holds it, so a direct SQL write cannot forge a
    verification without naming a real reviewer in the audit record.
    """
    try:
        result = verify_evidence(
            session,
            organization_id=principal.organization_id,
            reviewer_id=principal.user_id,
            evidence_id=evidence_id,
            confidence=payload.confidence,
        )
    except CompetitorError as exc:
        raise _refuse(exc) from exc
    session.commit()
    return result


@router.post(
    "/{competitor_product_id}/benchmarks",
    status_code=status.HTTP_201_CREATED,
    summary="Record a measured comparison against our own work",
)
def post_benchmark(
    competitor_product_id: uuid.UUID,
    payload: BenchmarkCreate,
    principal: Principal = Depends(
        require_permission("material.edit", "test.view", require_all=True)
    ),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """It cites a test; it does not grade one. Testing owns GREEN/YELLOW/RED.

    🔴 THIS WAS GATED ON `test.view` ALONE, WHICH IS A READ PERMISSION ON A
    WRITE ROUTE (Codex P1, 2026-08-28). Anybody who could merely LOOK at tests
    could author competitor comparisons and gap summaries, and RLS could not
    stop it: the writer is inside a project they legitimately reach, so the
    policy passes. A read permission never authorizes a write.

    BOTH, not either. `material.edit` is the authoring right the rest of this
    module uses; `test.view` is needed because a benchmark cites a test and
    displays values drawn from it, so an author who cannot see tests would be
    recording a comparison against something invisible to them.

    Measured: `product_development_chemist` holds both, which is exactly who
    benchmarks a competitor. `procurement_specialist` holds `material.edit`
    without `test.view` and is correctly excluded.
    """
    try:
        result = record_benchmark(
            session,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            competitor_product_id=competitor_product_id,
            project_id=payload.project_id,
            attribute=payload.attribute,
            gap_summary=payload.gap_summary,
            competitor_value=payload.competitor_value,
            our_value=payload.our_value,
            formula_version_id=payload.formula_version_id,
            test_id=payload.test_id,
        )
    except CompetitorError as exc:
        raise _refuse(exc) from exc
    session.commit()
    return result


@router.get("/{competitor_product_id}/benchmarks", summary="Measured comparisons on file")
def get_benchmarks(
    competitor_product_id: uuid.UUID,
    principal: Principal = Depends(require_permission("test.view")),
    session: Session = Depends(get_db),
) -> list[dict[str, Any]]:
    """⚠️ `test.view`, MATCHING ITS WRITER — a benchmark is testing output.

    A reader who may not see tests may not see comparisons drawn from them,
    which would otherwise be a way to read test results through a side door.
    """
    return list_benchmarks(
        session,
        organization_id=principal.organization_id,
        competitor_product_id=competitor_product_id,
    )


class AdoptPublicProduct(BaseModel):
    """Bring a public catalogue product into this tenant's pipeline."""

    public_product_id: uuid.UUID
    project_id: uuid.UUID | None = None


@router.post(
    "/from-public",
    status_code=status.HTTP_201_CREATED,
    summary="Adopt a product from the public catalogue into this tenant",
)
def post_adopt_public_product(
    payload: AdoptPublicProduct,
    principal: Principal = Depends(require_permission("material.edit")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """The signed-in half of a public product card.

    🔴 IT MAKES NO COMPOSITION CLAIM. It creates the tenant's own competitor
    record and links it to the public row; the evidence matrix stays empty
    until a document, a sample or a test fills it. This application does not
    hold a competitor's formula and does not infer one — 056 settled that.

    `material.edit`, the same permission `POST /competitors/products` already
    requires, because this IS that write with its fields read from the public
    catalogue rather than typed by hand.
    """
    try:
        result = adopt_public_product(
            session,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            public_product_id=payload.public_product_id,
            project_id=payload.project_id,
        )
    except CompetitorNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    except CompetitorError as exc:
        raise _refuse(exc) from exc
    session.commit()
    return result
